### PR TITLE
Add a backend for IonQ.

### PR DESCRIPTION
**Pull request by amilstead**
Hello ProjectQ!

Here is a first pass at a direct IonQ backend. Welcoming all feedback!

### CODE REVIEW

File: projectq/setups/ionq.py
Comment (Takishima) [inline projectq/setups/ionq.py:71]: Even though the mapper is not doing anything meaningful here, it should still be located at the end of the engine list and not at the beginning.

```suggestion
    return engine_list + [mapper]
```

**Review by Takishima — CHANGES_REQUESTED** [review]
Thank you very much for this contribution. I really have not much to say since everything is well-written and documented.

Apart from the two comments that I have made, could you make sure that the year of the license header is 2021 and not 2020?

**Comment by amilstead** [comment]
@Takishima Thanks for the review! I believe this is ready for another pass.

**Review by Takishima — APPROVED** [review]
Thanks for the latest changes and I am happy to merge the proposed changes.

I am just waiting on the confirmation that you have signed the ProjectQ license agreement (you should have gotten an email or will get it soon about the details of it). Once that is done I will merge the branch and prepare the release of a new version of ProjectQ as soon as possible.

**Comment by amilstead** [comment]
> I am just waiting on the confirmation that you have signed the ProjectQ license agreement (you should have gotten an email or will get it soon about the details of it). Once that is done I will merge the branch and prepare the release of a new version of ProjectQ as soon as possible.

Awesome! Thanks a lot! Looking forward to the release. 🎉

**Comment by Takishima** [comment]
I have received confirmation that the CLA was signed a few days ago. I will merge this in the coming days and start the release process for the next version of ProjectQ shortly thereafter.
However, I would like to have PR #395 land first (and it should be reviewed soon) since that one has a few important features pertaining to the release process.

**Comment by amilstead** [comment]
> I will merge this in the coming days and start the release process for the next version of ProjectQ shortly thereafter.

🎉 Awesome! Thank you!

> However, I would like to have PR #395 land first

Sounds good! I'll keep an eye on that PR. Please let me know if there is anything else I can do until the merge.

**Comment by amilstead** [comment]
Hey @Takishima -- any updates on when we might be able to expect #395 to merge?

Thanks!

**Comment by Takishima** [comment]
> Hey @Takishima -- any updates on when we might be able to expect #395 to merge?
> 
> Thanks!

I wish I could give you a satisfactory answer, but I am currently waiting on @andreashehn to finish his review (he has started looking at it). 

Unfortunately, he has been pretty busy with his work over the last few weeks. My hope is that by the end of this week or beginning of the next he should be able to find some time to finish his review and then I will be able to merge #395 and this PR.

**Comment by amilstead** [comment]
Thanks! I'll check back in early next week!

**Comment by amilstead** [comment]
Hey @Takishima,

Checking in as promised. Any luck with #395? I see no movement so far so assuming not, but checking nonetheless.

File: projectq/backends/_ionq/_ionq.py
Comment (Takishima) [inline projectq/backends/_ionq/_ionq.py:185]: This is a problem. Hard-coding a reset of the qubit index from the `MainEngine` here deep in some backend code is going to come and bite us back at some point.

Additionally, it is not uncommon to use a backend as a compiler engine (e.g. using a `CommandPrinter` before the simulator). If you arbitrarily reset the internal counter for the qubit IDs here, this has the potential to mess things up.

I would suggest that you remove this and handle the incrementing qubit IDs using a dedicated mapper (which you could automatically add to the `get_engine_list` function in the IonQ setup.

I don't think that we currently have a mapper that is able to do what you want, but I could imaging something along those lines (code not tested):

```python
from projectq.cengines import BasicMapperEngine
from projectq.meta import LogicalQubitIDTag
from projectq.ops import Command, AllocateQubitGate, DeallocateQubitGate
from projectq.types import WeakQubitRef


class BoundedQubitMapper(BasicMapperEngine):
    def __init__(self, n_qubits):
        super().__init__()
        self._qubit_idx = 0
        self.n_qubits = n_qubits

    def _reset(self):
        self._qubit_idx = 0

        # Also automatically send deallocate commands here maybe?

    def _process_cmd(self, cmd):
        if not self.current_mapping:
            self.current_mapping = dict()

        if isinstance(cmd.gate, AllocateQubitGate):
            qubit_id = cmd.qubits[0][0].id
            if qubit_id in self.current_mapping:
                raise RuntimeError("Cannot allocate an already allocated qubit")
            else:
                if self._qubit_idx == n_qubits - 1:
                    raise RuntimeError("Too many qubits")

                new_id = self._qubit_idx
                self._qubit_idx += 1

                self.current_mapping[qubit_id] = new_id

                qb = WeakQubitRef(engine=self, idx=new_id)
                new_cmd = Command(
                    engine=self,
                    gate=AllocateQubitGate(),
                    qubits=([qb],),
                    tags=[LogicalQubitIDTag(qubit_id)],
                )
                self.send([new_cmd])
        elif isinstance(cmd.gate, DeallocateQubitGate):
            qubit_id = cmd.qubits[0][0].id
            if qubit_id not in self.current_mapping:
                raise RuntimeError(
                    "Cannot deallocate a qubit that is not already allocated"
                )
            else:
                qb = WeakQubitRef(engine=self, idx=self.current_mapping[qubit_id])
                new_cmd = Command(
                    engine=self,
                    gate=DeallocateQubitGate(),
                    qubits=([qb],),
                    tags=[LogicalQubitIDTag(qubit_id)],
                )

                self._current_mapping.pop(qubit_id)
                self.send([new_cmd])
        else:
            self._send_cmd_with_mapped_ids(cmd)

    def receive(self, command_list):
        for cmd in command_list:
            if isinstance(cmd.gate, FlushGate):
                self._reset()
                self.send([cmd])
            else:
                self._process_cmd(cmd)
```

**Review by Takishima — CHANGES_REQUESTED** [review]
Again, thanks for your contributions and patience with me in this process that is taking way longer than I would like.

One of your latest changes has triggered some warning bells on my side (see comment in the code) and I would be happy to discuss this more. I made one suggestion but if you think you have a better way of solving the issue, I am open to discussing this here.

**Comment by Takishima** [comment]
> Hey @Takishima,
> 
> Checking in as promised. Any luck with #395? I see no movement so far so assuming not, but checking nonetheless.

No unfortunately not. I am gently reminding him every week without success so far, but I am sure that once things settle down a little for him at work he will submit his review.

File: projectq/backends/_ionq/_ionq.py
Comment (amilstead) [reply projectq/backends/_ionq/_ionq.py:185]: > This is a problem. Hard-coding a reset of the qubit index from the `MainEngine` here deep in some backend code is going to come and bite us back at some point.
> 
> Additionally, it is not uncommon to use a backend as a compiler engine (e.g. using a `CommandPrinter` before the simulator). If you arbitrarily reset the internal counter for the qubit IDs here, this has the potential to mess things up.
> 
> ```python
> eng = MainEngine(engine_list=[CommandPrinter(), IonQ(...)])
> ```
> 
> I would suggest that you remove this and handle the incrementing qubit IDs using a dedicated mapper (which you could automatically add to the `get_engine_list` function in the IonQ setup.
> 
> I don't think that we currently have a mapper that is able to do what you want, but I could imaging something along those lines (code not tested):
> 
> ```python
> from projectq.cengines import BasicMapperEngine
> from projectq.meta import LogicalQubitIDTag
> from projectq.ops import Command, AllocateQubitGate, DeallocateQubitGate
> from projectq.types import WeakQubitRef
> 
> 
> class BoundedQubitMapper(BasicMapperEngine):
>     def __init__(self, n_qubits):
>         super().__init__()
>         self._qubit_idx = 0
>         self.n_qubits = n_qubits
> 
>     def _reset(self):
>         self._qubit_idx = 0
> 
>         # Also automatically send deallocate commands here maybe?
> 
>     def _process_cmd(self, cmd):
>         if not self.current_mapping:
>             self.current_mapping = dict()
> 
>         if isinstance(cmd.gate, AllocateQubitGate):
>             qubit_id = cmd.qubits[0][0].id
>             if qubit_id in self.current_mapping:
>                 raise RuntimeError("Cannot allocate an already allocated qubit")
>             else:
>                 if self._qubit_idx == n_qubits - 1:
>                     raise RuntimeError("Too many qubits")
> 
>                 new_id = self._qubit_idx
>                 self._qubit_idx += 1
> 
>                 self.current_mapping[qubit_id] = new_id
> 
>                 qb = WeakQubitRef(engine=self, idx=new_id)
>                 new_cmd = Command(
>                     engine=self,
>                     gate=AllocateQubitGate(),
>                     qubits=([qb],),
>                     tags=[LogicalQubitIDTag(qubit_id)],
>                 )
>                 self.send([new_cmd])
>         elif isinstance(cmd.gate, DeallocateQubitGate):
>             qubit_id = cmd.qubits[0][0].id
>             if qubit_id not in self.current_mapping:
>                 raise RuntimeError(
>                     "Cannot deallocate a qubit that is not already allocated"
>                 )
>             else:
>                 qb = WeakQubitRef(engine=self, idx=self.current_mapping[qubit_id])
>                 new_cmd = Command(
>                     engine=self,
>                     gate=DeallocateQubitGate(),
>                     qubits=([qb],),
>                     tags=[LogicalQubitIDTag(qubit_id)],
>                 )
> 
>                 self._current_mapping.pop(qubit_id)
>                 self.send([new_cmd])
>         else:
>             self._send_cmd_with_mapped_ids(cmd)
> 
>     def receive(self, command_list):
>         for cmd in command_list:
>             if isinstance(cmd.gate, FlushGate):
>                 self._reset()
>                 self.send([cmd])
>             else:
>                 self._process_cmd(cmd)
> ```

Ahh, this is the missing piece to the puzzle! I was hoping this would come up. Setting that index felt _really_ awful, but I wasn't sure how else to support this. I did not realize that Mappers were able to process commands like this, so what you have provided here is way better. Thanks for this! 

Will plan to update this today!

**Comment by amilstead** [comment]
@Takishima thanks again for the nudge in the right direction with this qubit mapper. I am WAY happier with how this looks/feels now. 

Please let me know what still needs work!

File: projectq/backends/_ionq/_ionq_mapper.py
Comment (amilstead) [inline projectq/backends/_ionq/_ionq_mapper.py:13]: Luckily I didn't need to worry about forcibly deallocating-- the important part was just resetting the index and validating it is within bounds. The plain dict takes care of the rest.

**Review by Takishima — APPROVED** [review]
Looks good to me.

**Comment by Takishima** [comment]
So, #395 just landed and it is probably going to require a few modifications to your code, though I do not expect anything too major. You might want to rebase you branch onto the new develop and force-push it to update this PR.

**Comment by amilstead** [comment]
> So, #395 just landed and it is probably going to require a few modifications to your code, though I do not expect anything too major. You might want to rebase you branch onto the new develop and force-push it to update this PR.

Thanks! I'll work on this starting today.

**Comment by amilstead** [comment]
@Takishima This should be updated now. That said, I did run into some issues with `pytest` locally, on `revkit` code, but this happened on `develop` for me as well.

One other thing I decided to push here too-- making sure that ProjectQ qubit IDs are actually used when setting Qubit object result values from our API response.

**Review by Takishima — APPROVED** [review]

**Comment by Takishima** [comment]
I am happy to merge this if you are happy with it too ;-)

**Comment by amilstead** [comment]
> I am happy to merge this if you are happy with it too ;-)

Let's rock and roll!! Thanks for your patience and attention on this!

**Comment by Takishima** [comment]
> > I am happy to merge this if you are happy with it too ;-)
> 
> Let's rock and roll!! Thanks for your patience and attention on this!

Thank you for taking the time to adjust and tweak your code along the way. I will merge this later tonight and then probably release a new version of ProjectQ by the end of the week.

**Comment by amilstead** [comment]
> > > I am happy to merge this if you are happy with it too ;-)
> 
> > 
> 
> > Let's rock and roll!! Thanks for your patience and attention on this!
> 
> 
> 
> Thank you for taking the time to adjust and tweak your code along the way. I will merge this later tonight and then probably release a new version of ProjectQ by the end of the week.

Sounds great. Please let me know what I can do to help, if anything.

**Pull request merged**
